File: src/edictum_server/schemas/compositions.py
"""Request and response schemas for bundle composition endpoints."""

from __future__ import annotations

import re
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

_BUNDLE_NAME_RE = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
_BUNDLE_NAME_MAX_LEN = 128


class CompositionItemInput(BaseModel):
    """A contract reference within a composition."""

    contract_id: str = Field(..., max_length=128)  # stable contract_id (not UUID)
    position: int
    mode_override: Literal["enforce", "observe"] | None = None
    enabled: bool = True


class CompositionCreateRequest(BaseModel):
    """Create a new bundle composition."""

    name: str = Field(..., max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    defaults_mode: Literal["enforce", "observe"] = "enforce"
    update_strategy: Literal["manual", "auto_deploy", "observe_first"] = "manual"
    contracts: list[CompositionItemInput] = Field(default=[], max_length=100)
    tools_config: dict[str, Any] | None = None
    observability: dict[str, Any] | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        if len(v) > _BUNDLE_NAME_MAX_LEN:
            raise ValueError(f"name must be at most {_BUNDLE_NAME_MAX_LEN} characters")
        if not _BUNDLE_NAME_RE.match(v):
            raise ValueError(
                "name must match [a-z0-9][a-z0-9._-]* "
                "(lowercase, digits, dots, hyphens, underscores)"
            )
        return v


class CompositionUpdateRequest(BaseModel):
    """Update a bundle composition (all fields optional)."""

    description: str | None = Field(default=None, max_length=2000)
    defaults_mode: Literal["enforce", "observe"] | None = None
    update_strategy: Literal["manual", "auto_deploy", "observe_first"] | None = None
    contracts: list[CompositionItemInput] | None = Field(default=None, max_length=100)
    tools_config: dict[str, Any] | None = None
    observability: dict[str, Any] | None = None


class CompositionItemDetail(BaseModel):
    """Detailed view of a contract within a composition."""

    model_config = ConfigDict(from_attributes=True)

    contract_id: str  # stable id
    contract_name: str
    contract_type: str
    contract_version: int
    position: int
    mode_override: str | None
    enabled: bool
    has_newer_version: bool


class CompositionSummary(BaseModel):
    """Composition list item."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    description: str | None
    defaults_mode: str
    update_strategy: str
    contract_count: int
    updated_at: datetime
    created_by: str


class CompositionDetail(CompositionSummary):
    """Full composition with contracts and config."""

    id: uuid.UUID
    tenant_id: uuid.UUID
    contracts: list[CompositionItemDetail]
    tools_config: dict[str, Any] | None
    observability: dict[str, Any] | None


class PreviewResponse(BaseModel):
    """Assembled YAML preview (no deployment)."""

    yaml_content: str
    contracts_count: int
    validation_errors: list[str]


class ComposeDeployRequest(BaseModel):
    """Request body for deploying a composed bundle."""

    env: Literal["production", "staging", "development"]


class ComposeDeployResponse(BaseModel):
    """Response after assembling, signing, and deploying."""

    bundle_name: str
    bundle_version: int
    contracts_assembled: list[dict[str, Any]]
    deployment_id: uuid.UUID

File: dashboard/src/lib/api/compositions.ts
import { request, requestVoid } from "./client"

// --- Composition types ---

export interface CompositionSummary {
  name: string
  description: string | null
  defaults_mode: string
  update_strategy: string
  contract_count: number
  updated_at: string
  created_by: string
}

export interface CompositionItemDetail {
  contract_id: string
  contract_name: string
  contract_type: string
  contract_version: number
  position: number
  mode_override: string | null
  enabled: boolean
  has_newer_version: boolean
}

export interface CompositionDetail extends CompositionSummary {
  id: string
  tenant_id: string
  contracts: CompositionItemDetail[]
  tools_config: Record<string, unknown> | null
  observability: Record<string, unknown> | null
}

export interface PreviewResponse {
  yaml_content: string
  contracts_count: number
  validation_errors: string[]
}

export interface ComposeDeployResponse {
  bundle_name: string
  bundle_version: number
  contracts_assembled: Record<string, unknown>[]
  deployment_id: string
}

// --- Composition API ---

export function listCompositions(): Promise<CompositionSummary[]> {
  return request<CompositionSummary[]>("/compositions")
}

export function getComposition(name: string): Promise<CompositionDetail> {
  return request<CompositionDetail>(
    `/compositions/${encodeURIComponent(name)}`,
  )
}

export function createComposition(body: {
  name: string
  description?: string
  defaults_mode?: string
  update_strategy?: string
}): Promise<CompositionDetail> {
  return request<CompositionDetail>("/compositions", {
    method: "POST",
    body: JSON.stringify(body),
  })
}

export interface CompositionItemInput {
  contract_id: string
  position: number
  mode_override: "enforce" | "observe" | null
  enabled: boolean
}

export function updateComposition(
  name: string,
  body: {
    description?: string
    defaults_mode?: string
    update_strategy?: string
    contracts?: CompositionItemInput[]
  },
): Promise<CompositionDetail> {
  return request<CompositionDetail>(
    `/compositions/${encodeURIComponent(name)}`,
    { method: "PUT", body: JSON.stringify(body) },
  )
}

export function deleteComposition(name: string): Promise<void> {
  return requestVoid(`/compositions/${encodeURIComponent(name)}`, {
    method: "DELETE",
  })
}

export function previewComposition(name: string): Promise<PreviewResponse> {
  return request<PreviewResponse>(
    `/compositions/${encodeURIComponent(name)}/preview`,
    { method: "POST" },
  )
}

export function deployComposition(
  name: string,
  env: string,
): Promise<ComposeDeployResponse> {
  return request<ComposeDeployResponse>(
    `/compositions/${encodeURIComponent(name)}/deploy`,
    { method: "POST", body: JSON.stringify({ env }) },
  )
}
